docs/source/tutorials/chat.rst

```diff
@@ -0,0 +1,331 @@
+If you look at our :class:`~torchtune.data.Llama2ChatFormat` class, you'll notice that
+we don't include the :code:`<s>` and :code:`</s>` tokens. These are the beginning-of-sequence
+(BOS) and end-of-sequence (EOS) tokens that are represented differently in the tokenizer
+than the rest of the prompt template. Let's tokenize this example with the
+:class:`~torchtune.modules.tokenizers.SentencePieceTokenizer` used by LLaMA2 to see
+why.
+
+.. code-block:: python
+
+    from torchtune.modules.tokenizers import SentencePieceTokenizer
+
+    tokenizer = SentencePieceTokenizer("/tmp/Llama-2-7b-hf/tokenizer.model")
+    user_message = formatted_messages[0].content
+    tokens = tokenizer.encode(user_message, add_bos=True, add_eos=True)
+    print(tokens)
+    # [1, 518, 25580, 29962, 3532, 14816, 29903, 6778, ..., 2]
+
+We've added the BOS and EOS tokens when encoding our example text. This shows up
+as IDs 1 and 2. We can verify that these are our BOS and EOS tokens.
+
+.. code-block:: python
+
+    print(tokenizer.spm_model.piece_to_id("<s>"))
+    # 1
+    print(tokenizer.spm_model.piece_to_id("</s>"))
+    # 2
+
+The BOS and EOS tokens are what we call special tokens, because they have their own
+reserved token IDs. This means that they will index to their own individual vectors in
+the model's learnt embedding table. The rest of the prompt template tags, :code:`[INST]`
+and :code:`<<SYS>>` are tokenized as normal text and not their own IDs.
+
+.. code-block:: python
+
+    print(tokenizer.decode(518))
+    # '['
+    print(tokenizer.decode(25580))
+    # 'INST'
+    print(tokenizer.decode(29962))
+    # ']'
+    print(tokenizer.decode([3532, 14816, 29903, 6778]))
+    # '<<SYS>>'
+
+It's important to note that you should not place the special reserved tokens in your
+input prompts manually, as it will be treated as normal text and not as a special
+token.
+
+.. code-block:: python
+
+    print(tokenizer.encode("<s>", add_bos=False, add_eos=False))
+    # [529, 29879, 29958]
+
+Now let's take a look at LLaMA3's formatting to see how it's tokenized differently
+than LLaMA2.
+
+.. code-block:: python
+
+    from torchtune.modules.tokenizers import TikTokenTokenizer
+
+    tokenizer = TikTokenTokenizer("/tmp/Meta-Llama-3-8B/original/tokenizer.model")
+    messages = [Message.from_dict(msg) for msg in sample]
+    tokens, mask = tokenizer.tokenize_messages(messages)
+    print(tokenizer.decode(tokens))
+    # '<|start_header_id|>system<|end_header_id|>\n\nYou are a helpful, respectful,
+    # and honest assistant.<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWho
+    # are the most influential hip-hop artists of all time?<|eot_id|><|start_header_id|>
+    # assistant<|end_header_id|>\n\nHere is a list of some of the most influential hip-hop
+    # artists of all time: 2Pac, Rakim, N.W.A., Run-D.M.C., and Nas.<|eot_id|>'
+
+.. note::
+    We used the `tokenize_messages` API for LLaMA3, which is different than
```

Comment:
nit: tokenize_messages this is showing up as italics in the doc, think you want double backticks here